src/core/lib/iomgr/tcp_server_posix.cc

```diff
@@ -105,6 +136,100 @@ static grpc_error_handle tcp_server_create(grpc_closure* shutdown_complete,
+          acceptor->fd_index = -1;
+          if (!is_external) {
+            auto it = s->listen_fd_to_index_map.find(listener_fd);
+            if (it != s->listen_fd_to_index_map.end()) {
+              acceptor->port_index = std::get<0>(it->second);
+              acceptor->fd_index = std::get<1>(it->second);
+            }
+          } else {
+            // External connection handling.
+            grpc_resolved_address addr;
+            memset(&addr, 0, sizeof(addr));
+            addr.len = static_cast<socklen_t>(sizeof(struct sockaddr_storage));
+            // Get the fd of the socket connected to peer.
+            int fd = reinterpret_cast<grpc_event_engine::experimental::
+                                          PosixEndpointWithFdSupport*>(ep.get())
+                         ->GetWrappedFd();
+            if (getpeername(fd, reinterpret_cast<struct sockaddr*>(addr.addr),
+                            &(addr.len)) < 0) {
+              gpr_log(GPR_ERROR, "Failed getpeername: %s",
+                      grpc_core::StrError(errno).c_str());
+              close(fd);
+              return;
+            }
+            (void)grpc_set_socket_no_sigpipe_if_possible(fd);
+            auto addr_uri = grpc_sockaddr_to_uri(&addr);
+            if (!addr_uri.ok()) {
+              gpr_log(GPR_ERROR, "Invalid address: %s",
+                      addr_uri.status().ToString().c_str());
+              return;
+            }
+            if (grpc_tcp_trace.enabled()) {
+              gpr_log(GPR_INFO,
+                      "SERVER_CONNECT: incoming external connection: %s",
+                      addr_uri->c_str());
+            }
+          }
+          grpc_pollset* read_notifier_pollset =
+              (*(s->pollsets))[static_cast<size_t>(gpr_atm_no_barrier_fetch_add(
+                                   &s->next_pollset_to_assign, 1)) %
+                               s->pollsets->size()];
+          acceptor->external_connection = is_external;
+          acceptor->listener_fd = listener_fd;
+          grpc_byte_buffer* buf = nullptr;
+          if (pending_data != nullptr && pending_data->Length() > 0) {
+            buf = grpc_raw_byte_buffer_create(nullptr, 0);
+            grpc_slice_buffer_swap(&buf->data.raw.slice_buffer,
+                                   pending_data->c_slice_buffer());
+            pending_data->Clear();
+          }
+          acceptor->pending_data = buf;
+          s->on_accept_cb(s->on_accept_cb_arg,
+                          grpc_event_engine::experimental::
+                              grpc_event_engine_endpoint_create(std::move(ep)),
+                          read_notifier_pollset, acceptor);
+        };
+    auto on_shutdown_complete_cb =
+        grpc_event_engine::experimental::GrpcClosureToStatusCallback(
+            shutdown_complete);
+    auto listener =
+        reinterpret_cast<PosixEventEngineWithFdSupport*>(
+            grpc_event_engine::experimental::GetDefaultEventEngine().get())
```

Comment:
I do wonder if we'll see lifetime issues here, listeners outliving the default EventEngine instance, but we _shouldn't_ see it if this contract can be maintained in gRPC core.

https://github.com/grpc/grpc/blob/1c4b138322bbe5f6091092013b318aece2d3a55a/include/grpc/event_engine/event_engine.h#L360-L367

Reply:
I dont think this would be an issue with the default event engine at least. The listeners take a shared ptr to the event engine, so the event engine is alive as long as the listener is alive.